aten/src/ATen/native/cuda/DistanceKernel.cu

```diff
@@ -0,0 +1,217 @@
+    static __forceinline__ __device__ scalar_t backward(const scalar_t diff, const scalar_t grad, const scalar_t dist, const scalar_t p) { return dist == 0.0 ? 0 : sign(diff) * std::pow(std::abs(diff), p - 1) * grad / std::pow(dist, p - 1); }
+  };
+
+  // Two norm
+  struct two {
+    static __forceinline__ __device__ void inc(scalar_t& agg, const scalar_t diff, const scalar_t p) { agg += diff * diff; }
+    static __forceinline__ __device__ scalar_t finish(const scalar_t agg, const scalar_t p) { return device_sqrt<scalar_t>(agg); }
+    static __forceinline__ __device__ void agg(scalar_t& update, const scalar_t other) { update += other; }
+    static __forceinline__ __device__ scalar_t backward(const scalar_t diff, const scalar_t grad, const scalar_t dist, const scalar_t p) { return dist == 0.0 ? 0 : grad * diff / dist; }
+  };
+
+  // General p norm
+  struct p {
+    static __forceinline__ __device__ void inc(scalar_t& agg, const scalar_t diff, const scalar_t p) { agg += std::pow(diff, p); }
+    static __forceinline__ __device__ scalar_t finish(const scalar_t agg, const scalar_t p) { return std::pow(agg, static_cast<scalar_t>(1) / p); }
+    static __forceinline__ __device__ void agg(scalar_t& update, const scalar_t other) { update += other; }
+    static __forceinline__ __device__ scalar_t backward(const scalar_t diff, const scalar_t grad, const scalar_t dist, const scalar_t p) { return dist == 0.0 ? 0 : diff * std::pow(std::abs(diff), p - 2) * grad / std::pow(dist, p - 1); }
+  };
+
+  // Inf norm
+  struct inf {
+    static __forceinline__ __device__ void inc(scalar_t& agg, const scalar_t diff, const scalar_t p) { if (diff > agg) { agg = diff; } }
+    static __forceinline__ __device__ scalar_t finish(const scalar_t agg, const scalar_t p) { return agg; }
+    static __forceinline__ __device__ void agg(scalar_t& update, const scalar_t other) { if (other > update) { update = other; } }
+    static __forceinline__ __device__ scalar_t backward(const scalar_t diff, const scalar_t grad, const scalar_t dist, const scalar_t p) { return grad * sign(diff) * (std::abs(diff) == dist); }
+  };
+
+};
+
+template <typename scalar_t, typename F>
+__global__ static void pdist_kernel_cuda_impl(scalar_t * result, const scalar_t * self, const int64_t n, const int64_t m, const scalar_t p) {
+  const int k = blockIdx.x;
+  const int stride = blockDim.x;
+
+  float n2 = n - .5;
+  // The -1 accounts for floating point truncation issues
+  int64_t i = static_cast<int64_t>((n2 - device_sqrt<scalar_t>(n2 * n2 - 2 * k - 1)));
+  int64_t j = k - n * i + i * (i + 1) / 2 + i + 1;
+
+  const scalar_t * const start = self + i * m;
+  const scalar_t * const end = start + m;
+  const scalar_t * a = start + threadIdx.x;
+  const scalar_t * b = self + j * m + threadIdx.x;
+  scalar_t agg = 0.0;
+  for (; a < end; a += stride, b += stride) {
+    F::inc(agg, std::abs(*a - *b), p);
+  }
+  
+  // Reduce warps
+  for (int offset = warp_size / 2; offset > 0; offset /= 2) {
```

Comment:
Within the kernels you can use the built-in `warpSize`

Reply:
This becomes an issue when I want to reduce across warps. When I try to compile with shared memory size set to to the number of warps (or some function there of) I get the following compiler error: 

```
note: the value of variable "warpSize" here cannot be used as a constant
```

I could just hard code 32 here, or something similar and just assume that the number of warps is going to be less than that, which seems reasonable but a little eeehhhh.